styling docs (#463)
vtex/faststore: docs/styling.md

# Styling in Store Framework Jamstack

> 

Reviewer: ## Style with Gatsby themes

Gatsby themes are plugins that include a `gatsby-config.js` file and add pre-configured functionality, data sourcing, and/or UI code to Gatsby sites. This approach allows you to control the site styling through a decoupled package, with version control and autonomy for updates. It is also possible to have more than one theme!

> 

Reviewer: ## Gatsby themes on the Store Framework Jamstack

Styling in Store Framework Jamstack leverages the library [theme-ui](https://theme-ui.com/). By using this library we have a pre-defined way of creating themes that can be shared across components on the store.

> Theme UI already has a [gatsby plugin](https://www.gatsbyjs.com/docs/how-to/styling/theme-ui/) available. This plugin, however, was not attending our needs because we want to optimize how CSS is merged at runtime, and this plugin merges the CSS at build time.

Reviewer: I don't think that we need to say it here, justifying the choices. What about explaing your plugin? 
Reduce lines by using: 

"The Store Framework Jamstack  [custom plugin](https://github.com/vtex/faststore/tree/master/packages/gatsby-plugin-theme-ui) optimizes how CSS is merged at runtime."

So we've built a [custom plugin](https://github.com/vtex/faststore/tree/master/packages/gatsby-plugin-theme-ui
) to attend this need.
## How to style

> Using `@vtex/gatsby-plugin-theme-ui`, the path to shadow styles is to export an `index.js` file inside the folder `src/@vtex/gatsby-plugin-theme-ui`.

Reviewer: Using the `@vtex/gatsby-plugin-theme-ui`,  export an `index.js` file inside the folder `src/@vtex/gatsby-plugin-theme-ui`.

If you've started developing with our [template](https://github.com/vtex-sites/storecomponents.store/blob/master/src/%40vtex/gatsby-plugin-theme-ui/index.ts) this file should be already there.

> Since each store built using store framework jamstack is created on top of a pre-existing theme ui theme, every CSS customization should merge the new CSS with the pre-existing one.

Reviewer: Since each store built using Store Framework Jamstack is created on top of a pre-existing UI theme, every CSS customization should merge the new CSS into a pre-existing one.

Author: > should merge the new CSS into a pre-existing one
Actually it is not the idea to merge into the pre-existing one, but actually creating a new theme and merge the new CSS with the old CSS.

To do that, you use the `createTheme` function imported from `@vtex/store-ui`.

>ℹ️ `@vtex/store-ui` is our plugin that exports common ecommerce components

### Modifying or adding new styles to themes

> To style a pre-existing plugin, you use the gatsby functionality of [theme shadowing](https://www.gatsbyjs.com/docs/how-to/plugins-and-themes/shadowing/).

Reviewer: To style a pre-existing plugin, use the gatsby functionality, [theme shadowing](https://www.gatsbyjs.com/docs/how-to/plugins-and-themes/shadowing/).

Author: Nice

> 

Reviewer: ### Doing your first UI change

> Suppose you want to alter the color of the `searchSuggestions` add to cart button component from primary to secondary.

Reviewer: Let's alter the color of the `searchSuggestions` add to cart button component from `primary` to `secondary`.

 <img src="../images/suggestions.png">.

> You would need to find the correspondent for this button and alter the styles using theme shadowing. For the case of search suggestions, the theme for this component is located at this [file](https://github.com/vtex/faststore/blob/master/packages/gatsby-theme-store/src/components/SearchSuggestions/theme.ts).

Reviewer: First, find the correspondent for this button and alter the styles using theme shadowing.  

In your template, the theme for this component is located at this [file](https://github.com/vtex/faststore/blob/master/packages/gatsby-theme-store/src/components/SearchSuggestions/theme.ts).

```ts
// before altering the color
    products: {
      width: 'inherit',

      button: {
        backgroundColor: 'primary',
      },

      title,

      list: {
        ...list,
        display: 'flex',
        flexWrap: 'nowrap',
      },

      total: {
        paddingTop: '10px',
        color: 'text',
        textDecoration: 'underline',
        textAlign: 'center',
        cursor: 'pointer',
        width: '100%',
        backgroundColor: 'white',
      },
    },
```

> To alther the color of the button, you need to change the `backgroundColor` property from `primary` to `secondary`.

Reviewer: Then, alther the color of the button by changing the `backgroundColor` property from `primary` to `secondary`.

```
// after altering the color
    products: {
      width: 'inherit',

      button: {
        backgroundColor: 'secondary',
      },

      title,

      list: {
        ...list,
        display: 'flex',
        flexWrap: 'nowrap',
      },

      total: {
        paddingTop: '10px',
        color: 'text',
        textDecoration: 'underline',
        textAlign: 'center',
        cursor: 'pointer',
        width: '100%',
        backgroundColor: 'white',
      },
    },
```

## Remarks

> It is important to understand that theme shadowing only occurs on `@vtex/gatsby-plugin-theme-ui/index`. We structure the code with multiple files to improve readability. But the plugin only considers what is exported on  `index.ts`. Every theme from the components should be an argument for the `createTheme` function.

Reviewer: The theme shadowing only occurs on `@vtex/gatsby-plugin-theme-ui/index`. We structure the code with multiple files to improve readability, but the plugin only considers what is exported on  `index.ts`. Every theme from the components should be an argument for the `createTheme` function.

Example:

```ts
// @vtex/gatsby-plugin-theme-ui/index.ts
export default createTheme(
  base,
  breadcrumb,
  minicart,
  infoCardTheme,
  headerTheme,
  productQuantityTheme,
  minicartTheme,
  searchControlsTheme,
  searchSuggestionsTheme,
  searchBarTheme,
  searchTheme,
  custom,
  sliderTheme,
  offerTheme,
  productSummaryTheme,
  productDetailsTheme,
  loginTheme,
  authTheme
)
```

## Adding styles to a react component

> To add new styles from a theme into a react component, you have to discover what is the name of the variant that

Reviewer: Discover what is the name of the variant that uses the theme, than change it to de desired one.
 
 (Not sure this is it, but...)

# Styling in Store Framework Jamstack

Styling in Store Framework Jamstack leverages the library [theme-ui](https://theme-ui.com/). By using this library we have a pre-defined way of creating themes that can be shared across components on the store.

Theme UI already has a [gatsby plugin](https://www.gatsbyjs.com/docs/how-to/styling/theme-ui/) available. This plugin, however, was not attending our needs because we want to optimize how CSS is merged at runtime, and this plugin merges the CSS at build time.

So we've built a [custom plugin](https://github.com/vtex/faststore/tree/master/packages/gatsby-plugin-theme-ui

> ) to attend this need.

Reviewer 2: ```suggestion
) to attend to this need.
```

## How to style
Using `@vtex/gatsby-plugin-theme-ui`, the path to shadow styles is to export an `index.js` file inside the folder `src/@vtex/gatsby-plugin-theme-ui`.

If you've started developing with our [template](https://github.com/vtex-sites/storecomponents.store/blob/master/src/%40vtex/gatsby-plugin-theme-ui/index.ts) this file should be already there.

> Since each store built using store framework jamstack is created on top of a pre-existing theme ui theme, every CSS customization should merge the new CSS with the pre-existing one.

Reviewer 2: ```suggestion
Since each store built using store framework Jamstack is created on top of a pre-existing theme-ui theme, every CSS customization should merge the new CSS with the pre-existing ones.
```

> # Styling in Store Framework Jamstack
> 
> ## Style with Gatsby themes
> Gatsby themes are plugins that include a gatsby-config.js file and add pre-configured functionality, data sourcing, and/or UI code to Gatsby sites. This approach allows you to control the site styling through a decoupled package, with version control and autonomy for updates. It is also possible to have more than one theme!
> 
> ## Gatsby themes on the Store Framework Jamstack
> 
> Styling in Store Framework Jamstack leverages the library [theme-ui](https://theme-ui.com/). By using this library we have a pre-defined way of creating themes that can be shared across components on the store.
> 
> In order to optimize how CSS is merged at runtime, Store Framework Jamstack uses a custom `theme-ui` [plugin](https://github.com/vtex/faststore/tree/master/packages/gatsby-plugin-theme-ui)

Reviewer 3: ```suggestion
# SFJ Store Theme

The [SFJ Store Theme](https://github.com/vtex/faststore/tree/master/packages/gatsby-theme-store) is a [Gatsby theme](https://www.gatsbyjs.com/docs/themes/what-are-gatsby-themes/#gatsby-themes), a specific plugin type that contains pre-configured functionalities, data sourcing, and UI code defined in a [`/src/gatsby-config.js`](https://github.com/vtex/faststore/blob/master/packages/gatsby-theme-store/src/gatsby-config.ts) file.

Different from a plugin, in a Gatsby theme, default configurations (shared functionality, data sourcing, UI code) are abstracted into an installable package.

This way, Gatsby themes make it possible to control site styling through decoupled packages, characterized for version control and autonomous updates.

For being packaged as a standalone product, it is also possible to have more than one theme on the same site.

The SFJ Store Theme incorporates VTEX ecommerce solutions by configuring the following plugins:

- [`@vtex/gatsby-plugin-theme-ui`](https://github.com/vtex/faststore/tree/master/packages/gatsby-plugin-theme-ui) - Allows theme styling and optimizes the way CSS is merged at runtime. Leveraging from the [Theme UI](https://theme-ui.com/), this plugin incorporates the [Store UI library](https://github.com/vtex/faststore/tree/master/packages/store-ui) - a library that contains storefront components, such as the [Minicart](https://github.com/vtex/faststore/tree/master/packages/store-ui/src/Minicart).
- [`@vtex/gatsby-plugin-i18n`](https://github.com/vtex/faststore/tree/master/packages/gatsby-plugin-i18n) - Allows store internationalization.
- [`@vtex/gatsby-plugin-graphql`](https://github.com/vtex/faststore/tree/master/packages/gatsby-plugin-graphql) - Builds and preprocesses VTEX store's GraphQL queries.

# Styling

In the Store Framework Jamstack (SFJ), styling is achieved by [shadowing](https://www.gatsbyjs.com/docs/how-to/plugins-and-themes/shadowing/) the [`@vtex/gatsby-plugin-theme-ui`](https://github.com/vtex/faststore/tree/master/packages/gatsby-plugin-theme-ui).

>ℹ️ *[Shadowing](https://www.gatsbyjs.com/docs/conceptual/how-shadowing-works/) is a Gatsby concept that allows high levels of rendering customization by letting users override components, objects, or any other element from a theme’s `src` directory.*

Take the following steps as a guide to style any component from our [Store UI library](https://github.com/vtex/faststore/tree/master/packages/store-ui).

>ℹ️ *Leveraging from the [Theme UI](https://theme-ui.com/), the [`@vtex/gatsby-plugin-theme-ui`](https://github.com/vtex/faststore/tree/master/packages/gatsby-plugin-theme-ui) plugin incorporates the [Store UI library](https://github.com/vtex/faststore/tree/master/packages/store-ui) - a library that contains storefront components, such as the [Minicart](https://github.com/vtex/faststore/tree/master/packages/store-ui/src/Minicart).*
```

> # Styling in Store Framework Jamstack
> 
> ## Style with Gatsby themes
> Gatsby themes are plugins that include a gatsby-config.js file and add pre-configured functionality, data sourcing, and/or UI code to Gatsby sites. This approach allows you to control the site styling through a decoupled package, with version control and autonomy for updates. It is also possible to have more than one theme!
> 
> ## Gatsby themes on the Store Framework Jamstack
> 
> Styling in Store Framework Jamstack leverages the library [theme-ui](https://theme-ui.com/). By using this library we have a pre-defined way of creating themes that can be shared across components on the store.
> 
> In order to optimize how CSS is merged at runtime, Store Framework Jamstack uses a custom `theme-ui` [plugin](https://github.com/vtex/faststore/tree/master/packages/gatsby-plugin-theme-ui)

Reviewer 3: ```suggestion
# SFJ Store Theme

The [SFJ Store Theme](https://github.com/vtex/faststore/tree/master/packages/gatsby-theme-store) is a [Gatsby theme](https://www.gatsbyjs.com/docs/themes/what-are-gatsby-themes/#gatsby-themes), a specific plugin type that contains pre-configured functionalities, data sourcing, and UI code defined in a [`/src/gatsby-config.js`](https://github.com/vtex/faststore/blob/master/packages/gatsby-theme-store/src/gatsby-config.ts) file.

Different from a plugin, in a Gatsby theme, default configurations (shared functionality, data sourcing, UI code) are abstracted into an installable package.

This way, Gatsby themes make it possible to control site styling through decoupled packages, characterized for version control and autonomous updates.

For being packaged as a standalone product, it is also possible to have more than one theme on the same site.

The SFJ Store Theme incorporates VTEX ecommerce solutions by configuring the following plugins:

- [`@vtex/gatsby-plugin-theme-ui`](https://github.com/vtex/faststore/tree/master/packages/gatsby-plugin-theme-ui) - Allows theme styling and optimizes the way CSS is merged. Leveraging from the [Theme UI](https://theme-ui.com/), this plugin incorporates the [Store UI library](https://github.com/vtex/faststore/tree/master/packages/store-ui) - a library that contains storefront components, such as the [Minicart](https://github.com/vtex/faststore/tree/master/packages/store-ui/src/Minicart).
- [`@vtex/gatsby-plugin-i18n`](https://github.com/vtex/faststore/tree/master/packages/gatsby-plugin-i18n) - Allows store internationalization.
- [`@vtex/gatsby-plugin-graphql`](https://github.com/vtex/faststore/tree/master/packages/gatsby-plugin-graphql) - Builds and preprocesses VTEX store's GraphQL queries.

```

> 

Reviewer 3: ```suggestion
# Styling

In the Store Framework Jamstack (SFJ), styling can be achieved by:

- [Overwriting CSS styles of the Store UI components.](#overwriting-css-styles-of-the-store-ui-components)
- [Adding styles to React components from the Store UI library.](#adding-styles-to-react-components-from-the-store-ui-library)
- [Creating your own components and styles.](#creating-your-own-components-and-styles)

>ℹ️ *Leveraging from the [Theme UI](https://theme-ui.com/), the [`@vtex/gatsby-plugin-theme-ui`](https://github.com/vtex/faststore/tree/master/packages/gatsby-plugin-theme-ui) plugin incorporates the [Store UI library](https://github.com/vtex/faststore/tree/master/packages/store-ui) - a library that contains storefront components, such as the [Minicart](https://github.com/vtex/faststore/tree/master/packages/store-ui/src/Minicart).*

In the following sections, you'll find out more about each one of these topics.

However, before proceeding any further, keep in mind that styling is possible thanks to the `@vtex/gatsby-plugin-theme-ui` plugin. 

We structure the `@vtex/gatsby-plugin-theme-ui` code within multiple files to improve readability. However, the `@vtex/gatsby-plugin-theme-ui` plugin only considers the styles exported by the `createTheme` function inside the `index.ts` file.

    ```ts
    // @vtex/gatsby-plugin-theme-ui/index.ts
    export default createTheme(
      base,
      breadcrumb,
      minicart,
      infoCardTheme,
      headerTheme,
      productQuantityTheme,
      minicartTheme,
      searchControlsTheme,
      searchSuggestionsTheme,
      searchBarTheme,
      searchTheme,
      custom,
      sliderTheme,
      offerTheme,
      productSummaryTheme,
      productDetailsTheme,
      loginTheme,
      authTheme
    )
    ```
```

If you've started developing with our [template](https://github.com/vtex-sites/storecomponents.store/blob/master/src/%40vtex/gatsby-plugin-theme-ui/index.ts) this file should be already there.

Since each store built using Store Framework Jamstack is created on top of a pre-existing theme-ui theme, every CSS customization should merge the new CSS with the pre-existing ones.

To do that, you use the `createTheme` function imported from `@vtex/store-ui`.

>ℹ️ `@vtex/store-ui` is our plugin that exports common ecommerce components

### Doing your first CSS change
To style a pre-existing plugin, use the gatsby functionality [theme shadowing](https://www.gatsbyjs.com/docs/how-to/plugins-and-themes/shadowing/).

Let's alter the color of the `searchSuggestions` add to cart button component from primary to secondary.

 <img src="./images/suggestions.png">.

First, find the correspondent for this button and alter the styles using theme shadowing.

In your template, the theme for this component is located at this [file](https://github.com/vtex/faststore/blob/master/packages/gatsby-theme-store/src/components/SearchSuggestions/theme.ts).

```ts

> // before altering the color
>     products: {
>       width: 'inherit',
> 
>       button: {
>         backgroundColor: 'primary',
>       },
> 
>       title,
> 
>       list: {
>         ...list,
>         display: 'flex',
>         flexWrap: 'nowrap',
>       },
> 
>       total: {
>         paddingTop: '10px',
>         color: 'text',
>         textDecoration: 'underline',
>         textAlign: 'center',
>         cursor: 'pointer',
>         width: '100%',
>         backgroundColor: 'white',
>       },
>     },
> ```
> 
> Then, alther the color of the button by changing the `backgroundColor` property from `primary` to `secondary`.
> 
> 
> ```
> // after altering the color
>     products: {
>       width: 'inherit',
> 
>       button: {
>         backgroundColor: 'secondary',
>       },
> 
>       title,
> 
>       list: {
>         ...list,
>         display: 'flex',
>         flexWrap: 'nowrap',
>       },
> 
>       total: {
>         paddingTop: '10px',
>         color: 'text',
>         textDecoration: 'underline',
>         textAlign: 'center',
>         cursor: 'pointer',
>         width: '100%',
>         backgroundColor: 'white',
>       },
>     },
> ```
> <img src="./images/suggestions-altered-color.png">

Reviewer 3: ```suggestion
## Step by step

### Overwriting CSS styles of the Store UI components

Take the following steps as a guide to style Store UI components containing a pre-defined `themes.ts` style sheet.

Once you [set up your SFJ project]() from our [SFJ starter](https://github.com/vtex-sites/storecomponents.store), you're ready to start styling your VTEX store.  

1. Open up your SFJ project in the code editor of your choosing.
2. Go to the `src/@vtex/gatsby-plugin-theme-ui`.
3. Create a Typescript file with the name of the component you want to style (e.g., `searchSuggestions.ts`).

>ℹ️ *Use the [SFJ Store Theme components folder](https://github.com/vtex/faststore/tree/master/packages/gatsby-theme-store/src/components) as a reference.*

4. Import the `createTheme` function and the pre-defined style sheet of the component you want to style.
    
    ```ts
    import { createTheme, SxStyleProp } from '@vtex/store-ui'
    ```
    
>⚠️ *Ideally, all default style sheets should be defined within the Store UI library. However, you'll notice some styles are defined within the SFJ Store Theme. We are aware of this situation and already working on it.*

5. Declare a `custom` object, as in the following:

    ```ts
    const custom = {
    
    }
    ``` 

6. Within the `custom` object, overwrite styles respecting the hierarchy of the style sheet of that component.

>⚠️ *Open the default style sheet file of the component you want to style and check the hierarchy you should follow when overwriting styles.*

Take the following example from the [SearchSuggestions default style sheet (`@vtex/gatsby-theme-store/src/components/SearchSuggestions/theme.ts`)](https://github.com/vtex/faststore/blob/master/packages/gatsby-theme-store/src/components/SearchSuggestions/theme.ts). By default, we have:
    
    ```ts
    // "faststore/packages/gatsby-theme-store/src/components/SearchSuggestions/theme.ts"
        products: {
          width: 'inherit',
    
          button: {
            backgroundColor: 'primary',
          },
    
          title,
    
          list: {
            ...list,
            display: 'flex',
            flexWrap: 'nowrap',
          },
    
          total: {
            paddingTop: '10px',
            color: 'text',
            textDecoration: 'underline',
            textAlign: 'center',
            cursor: 'pointer',
            width: '100%',
            backgroundColor: 'white',
          },
        },
    ```

<img src="./images/suggestions.png">

>⚠️ ***Keep in mind**: If you don't respect the component style sheet hierarchy and names, you'll face issues when overwriting styles.*

To customize this style, we must stick to this hierarchy. Take the following example in which we customized the add-to-cart button from the `searchSuggestions` component by changing the `backgroundColor` property from `primary` to `secondary`.

<img src="./images/suggestions-altered-color.png">

    ```ts
    // "src/@vtex/gatsby-plugin-theme-ui/searchSuggestions.ts"
    import { createTheme, SxStyleProp } from '@vtex/store-ui'
    import searchSuggestionsTheme from '@vtex/gatsby-theme-store/components/SearchSuggestions/theme'
    
        const custom{
          button: {
            backgroundColor: 'secondary',
          }
        }
    ```

7. At the end of the file, create a `searchSuggestionsCustom` variable that provides the default theme and the `custom` theme you created to the `createTheme` function.

    ```ts
    const searchSuggestionsCustom = createTheme(searchSuggestionsTheme, custom)
    ```

8. Export the object created in the previous step.

    ```ts
    export default searchSuggestionsCustom
    ```

9. Save your changes and close this file.
10. Go to the `src/@vtex/gatsby-plugin-theme-ui/index.js` file.
11. Import the style you created in the previous steps . Take the following example:

    ```js
    import { searchSuggestionsCustom } from './searchSuggestions'
    ```
    
12. Export the theme in the `createTheme` function.

    
    ```js
    export default createTheme(
      searchSuggestionsCustom,
      ...
    )
    ```

7. Save your changes.

>ℹ️ ***Keep in mind**: [Shadowing]() only occurs on the `@vtex/gatsby-plugin-theme-ui/index` file.*

8. Run `gatsby deploy` to start a development server and check your changes live at `http://localhost:8000/`.

```

> ## Remarks
> Theme shadowing only occurs on `@vtex/gatsby-plugin-theme-ui/index`. We structure the code with multiple files to improve readability, but the plugin only considers what is exported on `index.ts`. Every theme from the components should be an argument for the `createTheme` function.
> 
> Example:
> 
> ```ts
> // @vtex/gatsby-plugin-theme-ui/index.ts
> export default createTheme(
>   base,
>   breadcrumb,
>   minicart,
>   infoCardTheme,
>   headerTheme,
>   productQuantityTheme,
>   minicartTheme,
>   searchControlsTheme,
>   searchSuggestionsTheme,
>   searchBarTheme,
>   searchTheme,
>   custom,
>   sliderTheme,
>   offerTheme,
>   productSummaryTheme,
>   productDetailsTheme,
>   loginTheme,
>   authTheme
> )
> ```

Reviewer 3: ```suggestion
```

> ## Adding styles to a react component
> To add new styles from a theme into a react component, discover what is the name of the Theme UI variant that you want to use and pass it as props to the component that you want to style. More about this on this theme-ui [doc](https://theme-ui.com/guides/variants/).

Reviewer 3: ```suggestion
## Adding styles to React components from the Store UI library

Leveraging from the [Theme UI](https://theme-ui.com/), the `@vtex/gatsby-plugin-theme-ui` plugin incorporates the [Store UI library](https://github.com/vtex/faststore/tree/master/packages/store-ui) - a library that contains storefront components, such as the [Minicart](https://github.com/vtex/faststore/tree/master/packages/store-ui/src/Minicart).

In this sense, we can benefit from the [Variants](https://theme-ui.com/guides/variants/) feature from the Theme UI library.

>ℹ️ *Variants allow you to add groups of styles based on a component value. It works as classes in CSS.*

Take the following example from the [`Header`](https://github.com/vtex/faststore/blob/master/packages/gatsby-theme-store/src/components/Header.tsx) component of the Store UI library:

    ```tsx
     const variant = 'header'

    ...

      <Flex variant={`${variant}.left`}>
        <Logo variant={`${variant}.logo`}/>
        <Menu variant={`${variant}.menu`}/>
      </Flex>
    
    ...

    ```

Since, in this case, the `variant` name is `header`, we should add styles to this component as in the following:

    ```ts
    {
      header:{
    	  left: {
    	    color: 'white',
          bg: 'secondary',
    	  },
    	   logo: {
            text: '#000',
            background: '#fff',
            primary: '#0c7',
         },
        menu{
            body: 'system-ui, sans-serif',
            heading: 'Poppins, sans-serif',
        }
    }
    ```

Therefore, to add new styles to a Store UI component, discover the name of its variants and pass it as props, containing the desired styles.

1. Open up your SFJ project in the code editor of your choosing.
2. Go to the `src/@vtex/gatsby-plugin-theme-ui/index.js` file.
3. Import the style of the component you want to style. 

>ℹ️ *Use the [SFJ Store UI library](https://github.com/vtex/faststore/tree/master/packages/store-ui) as a reference.*

Take the following example:

    ```js
    import headerTheme from '@vtex/store-ui'
    ```
    
4. Define styles inside the `src/@vtex/gatsby-plugin-theme-ui/index.js` file.
5. Save your changes.
6. Run `gatsby deploy` to start a development server and check your changes live at `http://localhost:8000/`.
```

        <Logo variant={`${variant}.logo`}/>
        <Menu variant={`${variant}.menu`}/>
      </Flex>
    
    ...

    ```

Since, in this case, the `variant` name is `header`, we should add styles to this component as in the following:

    ```ts
    {
      header:{
    	  left: {
    	    color: 'white',
          bg: 'secondary',
    	  },
    	   logo: {
            text: '#000',
            background: '#fff',
            primary: '#0c7',
         },
        menu{
            body: 'system-ui, sans-serif',
            heading: 'Poppins, sans-serif',
        }
    }
    ```

Therefore, to add new styles to a Store UI component, check the name of its variants inside the `.tsx` file. Then, pass the variants as props, containing the desired styles.

1. Open up your SFJ project in the code editor of your choosing.
2. Go to the `src/@vtex/gatsby-plugin-theme-ui/index.js` file.
3. Import the style of the component you want to style. 

>ℹ️ *Use the [SFJ Store UI library](https://github.com/vtex/faststore/tree/master/packages/store-ui) as a reference.*

Take the following example:

    ```js
    import headerTheme from '@vtex/store-ui'
    ```
    

> 4. Define styles inside the `src/@vtex/gatsby-plugin-theme-ui/index.js` file.

Reviewer 3: ```suggestion
4. Define styles inside the `src/@vtex/gatsby-plugin-theme-ui/index.js` file.

>ℹ️ *Check the name of the variants inside the component's `.tsx` source file. Then, pass the variants as props with the desired styles.*

```

5. Save your changes.

> 6. Run `gatsby deploy` to start a development server and check your changes live at `http://localhost:8000/`.

Reviewer 3: ```suggestion
6. Run `gatsby deploy` to start a development server and check your changes live at `http://localhost:8000/`.

## Creating your own components and styles

With SFJ, you don't need to feel stuck to pre-defined components.

By taking the following steps, you'll be able to create your own components:

1. Open up your SFJ project in the code editor of your choosing.
2. Go to the `src` folder.
3. Create a new folder with the name of your component.
4. Develop your React component and define styles.
5. Save your changes.
6. Now, go to the `src/@vtex/gatsby-plugin-theme-ui/index.js` file.
7. Import your component styles.

    ```js
    import {component-name} from '.../{component-folder}/{component}'
    ```

6. Save your changes.
7. Run `gatsby deploy` to start a development server and check your changes live at `http://localhost:8000/`.
```

Author: yarn develop não gatsby deploy

- [`@vtex/gatsby-plugin-i18n`](https://github.com/vtex/faststore/tree/master/packages/gatsby-plugin-i18n) - Allows store internationalization.
- [`@vtex/gatsby-plugin-graphql`](https://github.com/vtex/faststore/tree/master/packages/gatsby-plugin-graphql) - Builds and preprocesses VTEX store's GraphQL queries.

# Styling

In the Store Framework Jamstack (SFJ), styling can be achieved by:

- [Overwriting CSS styles of the Store UI components.](#overwriting-css-styles-of-the-store-ui-components)
- [Adding styles to React components from the Store UI library.](#adding-styles-to-react-components-from-the-store-ui-library)
- [Creating your own components and styles.](#creating-your-own-components-and-styles)

In the following sections, you'll find out more about each one of these topics.

However, before proceeding any further, keep in mind that styling is possible thanks to the `@vtex/gatsby-plugin-theme-ui` plugin. 

We structure the `@vtex/gatsby-plugin-theme-ui` code within multiple files to improve readability. However, the `@vtex/gatsby-plugin-theme-ui` plugin only considers the styles exported by the `createTheme` function inside the `index.ts` file.

>     ```ts

Reviewer 4: ![Screen Shot 2021-01-14 at 11 40 00 PM](https://user-images.githubusercontent.com/13058968/104674662-c4c7ba80-56c2-11eb-8ddc-0cb46fd940ec.png)

It's escaping somewhere?

    // @vtex/gatsby-plugin-theme-ui/index.ts
    export default createTheme(
      base,
      breadcrumb,
      minicart,
      infoCardTheme,
      headerTheme,
      productQuantityTheme,
      minicartTheme,
      searchControlsTheme,
      searchSuggestionsTheme,
      searchBarTheme,
      searchTheme,
      custom,
      sliderTheme,
      offerTheme,
      productSummaryTheme,
      productDetailsTheme,
      loginTheme,
      authTheme
    )
    ```
## Step by step

### Overwriting CSS styles of the Store UI components

Take the following steps as a guide to style Store UI components containing a pre-defined `themes.ts` style sheet.

Once you [set up your SFJ project]() from our [SFJ starter](https://github.com/vtex-sites/storecomponents.store), you're ready to start styling your VTEX store.  

1. Open up your SFJ project in the code editor of your choosing.
2. Go to the `src/@vtex/gatsby-plugin-theme-ui`.
3. Create a Typescript file with the name of the component you want to style (e.g., `searchSuggestions.ts`).

> >ℹ️ *Use the [SFJ Store Theme components folder](https://github.com/vtex/faststore/tree/master/packages/gatsby-theme-store/src/components) as a reference.*

Reviewer 4: ```suggestion
>ℹ️  *Use the [SFJ Store Theme components folder](https://github.com/vtex/faststore/tree/master/packages/gatsby-theme-store/src/components) as a reference.*
```

4. Import the `createTheme` function and the pre-defined style sheet of the component you want to style.
    
    ```ts
    import { createTheme, SxStyleProp } from '@vtex/store-ui'
    ```
    

> >⚠️ *Ideally, all default style sheets should be defined within the Store UI library. However, you'll notice some styles are defined within the SFJ Store Theme. We are aware of this situation and already working on it.*

Reviewer 4: ```suggestion
>⚠️  *Ideally, all default style sheets should be defined within the Store UI library. However, you'll notice some styles are defined within the SFJ Store Theme. We are aware of this situation and already working on it.*
```

5. Declare a `custom` object, as in the following:

    ```ts
    const custom = {
    
    }
    ``` 

6. Within the `custom` object, overwrite styles respecting the hierarchy of the style sheet of that component.

> >⚠️ *Open the default style sheet file of the component you want to style and check the hierarchy you should follow when overwriting styles.*

Reviewer 4: ```suggestion
>⚠️  *Open the default style sheet file of the component you want to style and check the hierarchy you should follow when overwriting styles.*
```

Take the following example from the [SearchSuggestions default style sheet (`@vtex/gatsby-theme-store/src/components/SearchSuggestions/theme.ts`)](https://github.com/vtex/faststore/blob/master/packages/gatsby-theme-store/src/components/SearchSuggestions/theme.ts). By default, we have:
    
    ```ts
    // "faststore/packages/gatsby-theme-store/src/components/SearchSuggestions/theme.ts"
        products: {
          width: 'inherit',
    
          button: {
            backgroundColor: 'primary',
          },
    
          title,
    
          list: {
            ...list,
            display: 'flex',
            flexWrap: 'nowrap',
          },
    
          total: {
            paddingTop: '10px',
            color: 'text',
            textDecoration: 'underline',
            textAlign: 'center',
            cursor: 'pointer',
            width: '100%',
            backgroundColor: 'white',
          },
        },
    ```

> <img src="./images/suggestions.png">

Reviewer 4: Can we define a width and a height here? The images look a little big [here](https://github.com/vtex/faststore/blob/docs/styling/docs/styling.md)

Author: Done

> >⚠️ ***Keep in mind**: If you don't respect the component style sheet hierarchy and names, you'll face issues when overwriting styles.*

Reviewer 4: ```suggestion
>⚠️  ***Keep in mind**: If you don't respect the component style sheet hierarchy and names, you'll face issues when overwriting styles.*
```

    ```ts
    // "src/@vtex/gatsby-plugin-theme-ui/searchSuggestions.ts"
    import { createTheme, SxStyleProp } from '@vtex/store-ui'
    import searchSuggestionsTheme from '@vtex/gatsby-theme-store/components/SearchSuggestions/theme'
    
        const custom{
          button: {
            backgroundColor: 'secondary',
          }
        }
    ```

7. At the end of the file, create a `searchSuggestionsCustom` variable that provides the default theme and the `custom` theme you created to the `createTheme` function.

    ```ts
    const searchSuggestionsCustom = createTheme(searchSuggestionsTheme, custom)
    ```

8. Export the object created in the previous step.

    ```ts
    export default searchSuggestionsCustom
    ```

9. Save your changes and close this file.
10. Go to the `src/@vtex/gatsby-plugin-theme-ui/index.js` file.
11. Import the style you created in the previous steps . Take the following example:

    ```js
    import searchSuggestionsCustom from './searchSuggestions'
    ```
    
12. Export the theme in the `createTheme` function.

    
    ```js
    export default createTheme(
      searchSuggestionsCustom,
      ...
    )
    ```

7. Save your changes.

> >ℹ️ ***Keep in mind**: [Shadowing]() only occurs on the `@vtex/gatsby-plugin-theme-ui/index` file.*

Reviewer 4: ```suggestion
>ℹ️  ***Keep in mind**: [Shadowing]() only occurs on the `@vtex/gatsby-plugin-theme-ui/index` file.*
```

> >ℹ️ ***Keep in mind**: [Shadowing]() only occurs on the `@vtex/gatsby-plugin-theme-ui/index` file.*

Reviewer 4: ```suggestion
>ℹ️  ***Keep in mind**: [Shadowing]() only occurs on the `@vtex/gatsby-plugin-theme-ui/index` file.*
```

8. Run `gatsby deploy` to start a development server and check your changes live at `http://localhost:8000/`.

## Adding styles to React components from the Store UI library

Leveraging from the [Theme UI](https://theme-ui.com/), the `@vtex/gatsby-plugin-theme-ui` plugin incorporates the [Store UI library](https://github.com/vtex/faststore/tree/master/packages/store-ui) - a library that contains storefront components, such as the [Minicart](https://github.com/vtex/faststore/tree/master/packages/store-ui/src/Minicart).

In this sense, we can benefit from the [Variants](https://theme-ui.com/guides/variants/) feature from the Theme UI library.

> >ℹ️ *Variants allow you to add groups of styles based on a component value. It works as classes in CSS.*

Reviewer 4: ```suggestion
>ℹ️  *Variants allow you to add groups of styles based on a component value. It works as classes in CSS.*
```

> >ℹ️ *Variants allow you to add groups of styles based on a component value. It works as classes in CSS.*

Reviewer 4: ```suggestion
>ℹ️  *Variants allow you to add groups of styles based on a component value. It works as classes in CSS.*
```

    ```tsx
     const variant = 'header'

    ...

      <Flex variant={`${variant}.left`}>
        <Logo variant={`${variant}.logo`}/>
        <Menu variant={`${variant}.menu`}/>
      </Flex>
    
    ...

    ```

Since, in this case, the `variant` name is `header`, we should add styles to this component as in the following:

    ```ts
    {
      header:{
    	  left: {
    	    color: 'white',
          bg: 'secondary',
    	  },
    	   logo: {
            text: '#000',
            background: '#fff',
            primary: '#0c7',
         },
        menu{
            body: 'system-ui, sans-serif',
            heading: 'Poppins, sans-serif',
        }
    }
    ```

Therefore, to add new styles to a Store UI component, check the name of its variants inside the `.tsx` file. Then, pass the variants as props, containing the desired styles.

1. Open up your SFJ project in the code editor of your choosing.
2. Go to the `src/@vtex/gatsby-plugin-theme-ui/index.js` file.
3. Import the style of the component you want to style. 

> >ℹ️ *Use the [SFJ Store UI library](https://github.com/vtex/faststore/tree/master/packages/store-ui) as a reference.*

Reviewer 4: ```suggestion
>ℹ️  *Use the [SFJ Store UI library](https://github.com/vtex/faststore/tree/master/packages/store-ui) as a reference.*
```

> >ℹ️ *Use the [SFJ Store UI library](https://github.com/vtex/faststore/tree/master/packages/store-ui) as a reference.*

Reviewer 4: ```suggestion
>ℹ️  *Use the [SFJ Store UI library](https://github.com/vtex/faststore/tree/master/packages/store-ui) as a reference.*
```

Take the following example:

    ```js
    import headerTheme from '@vtex/store-ui'
    ```
    
4. Define styles inside the `src/@vtex/gatsby-plugin-theme-ui/index.js` file.

> >ℹ️ *Check the name of the variants inside the component's `.tsx` source file. Then, pass the variants as props with the desired styles.*

Reviewer 4: ```suggestion
>ℹ️  *Check the name of the variants inside the component's `.tsx` source file. Then, pass the variants as props with the desired styles.*
```

> >ℹ️ *Check the name of the variants inside the component's `.tsx` source file. Then, pass the variants as props with the desired styles.*

Reviewer 4: ```suggestion
>ℹ️  *Check the name of the variants inside the component's `.tsx` source file. Then, pass the variants as props with the desired styles.*
```

- [Creating your own components and styles.](#creating-your-own-components-and-styles)

In the following sections, you'll find out more about each one of these topics.

However, before proceeding any further, keep in mind that styling is possible thanks to the `@vtex/gatsby-plugin-theme-ui` plugin. 

We structure the `@vtex/gatsby-plugin-theme-ui` code within multiple files to improve readability. However, the `@vtex/gatsby-plugin-theme-ui` plugin only considers the styles exported by the `createTheme` function inside the `index.ts` file.

```ts
// @vtex/gatsby-plugin-theme-ui/index.ts
export default createTheme(
    base,
    breadcrumb,
    minicart,
    infoCardTheme,
    headerTheme,
    productQuantityTheme,
    minicartTheme,
    searchControlsTheme,
    searchSuggestionsTheme,
    searchBarTheme,
    searchTheme,
    custom,
    sliderTheme,
    offerTheme,
    productSummaryTheme,
    productDetailsTheme,
    loginTheme,
    authTheme
)
```

## Step by step

### Overwriting CSS styles of the Store UI components

Take the following steps as a guide to style Store UI components containing a pre-defined `themes.ts` style sheet.

> Once you [set up your SFJ project]() from our [SFJ starter](https://github.com/vtex-sites/storecomponents.store), you're ready to start styling your VTEX store.  

Reviewer 3: ```suggestion
Once you [set up your SFJ project](getting-started.md) from our [SFJ starter](https://github.com/vtex-sites/storecomponents.store), you're ready to start styling your VTEX store.  
```
